sbkeys/generate-aws-sbkeys

```diff
@@ -0,0 +1,344 @@
+
+# Helper script for running commands to generate Secure Boot files.
+
+set -euo pipefail
+
+usage() {
+   cat >&2 <<EOF
+usage: ${0##*/} [--sdk-image SDK_IMAGE]
+                [--aws-region AWS_REGION]
+                [--pk-ca PK_CA]
+                [--kek-ca KEK_CA]
+                [--db-ca DB_CA]
+                [--vendor-ca VENDOR_CA]
+                [--shim-sign-key SHIM_SIGN_KEY]
+                [--code-sign-key CODE_SIGN_KEY]
+                [--config-sign-key CONFIG_SIGN_KEY]
+                [--output-dir OUTPUT_DIR]
+
+Generate Secure Boot related files. AWS-aware edition.
+
+Options:
+    --sdk-image         Name of the (optional) SDK image to use.
+    --aws-region        AWS region where the resources live.
+    --pk-ca             PCA ARN for the Secure Boot Platform CA (PK).
+    --kek-ca            PCA ARN for the Secure Boot Key Exchange CA (KEK).
+    --db-ca             PCA ARN for the Secure Boot Database CA (db).
+    --vendor-ca         PCA ARN for the Secure Boot Vendor CA.
+    --shim-sign-key     KMS key ID or ARN for the shim signing key.
+    --code-sign-key     KMS key ID or ARN for the code signing key (grub, vmlinuz).
+    --config-sign-key   KMS key ID or ARN for the config signing key (grub.cfg).
+    --output-dir        Path where the keys will be written.
+    --help              shows this usage text
+EOF
+}
+
+required_arg() {
+   local arg="${1:?}"
+   local value="${2}"
+   if [ -z "${value}" ]; then
+      echo "ERROR: ${arg} is required" >&2
+      exit 2
+   fi
+}
+
+parse_args() {
+  while [ ${#} -gt 0 ] ; do
+    case "${1}" in
+        --help ) usage; exit 0 ;;
+        --sdk-image ) shift; SDK_IMAGE="${1}" ;;
+        --aws-region ) shift; AWS_REGION="${1}" ;;
+        --pk-ca ) shift; PK_CA="${1}" ;;
+        --kek-ca ) shift; KEK_CA="${1}" ;;
+        --db-ca ) shift; DB_CA="${1}" ;;
+        --vendor-ca ) shift; VENDOR_CA="${1}" ;;
+        --shim-sign-key ) shift; SHIM_SIGN_KEY="${1}" ;;
+        --code-sign-key ) shift; CODE_SIGN_KEY="${1}" ;;
+        --config-sign-key ) shift; CONFIG_SIGN_KEY="${1}" ;;
+        --output-dir ) shift; OUTPUT_DIR="${1}" ;;
+        *) ;;
+    esac
+    shift
+  done
+
+  # Required arguments
+  required_arg "--aws-region" "${AWS_REGION:-}"
+  required_arg "--pk-ca" "${PK_CA:-}"
+  required_arg "--kek-ca" "${KEK_CA:-}"
+  required_arg "--db-ca" "${DB_CA:-}"
+  required_arg "--vendor-ca" "${VENDOR_CA:-}"
+  required_arg "--shim-sign-key" "${SHIM_SIGN_KEY:-}"
+  required_arg "--code-sign-key" "${CODE_SIGN_KEY:-}"
+  required_arg "--config-sign-key" "${CONFIG_SIGN_KEY:-}"
+  required_arg "--output-dir" "${OUTPUT_DIR:-}"
+}
+
+parse_args "${@}"
+
+PRELUDE_END=$(awk '/=\^\.\.\^=/ { print NR+1; exit 0; }' "${0}")
+SBKEYS_SCRIPT="$(mktemp)"
+AWS_KMS_PKCS11_CONF="$(mktemp)"
+cleanup() {
+  rm -f "${SBKEYS_SCRIPT}" "${AWS_KMS_PKCS11_CONF}"
+}
+trap 'cleanup' EXIT
+tail -n +"${PRELUDE_END}" "${0}" >"${SBKEYS_SCRIPT}"
+chmod +x "${SBKEYS_SCRIPT}"
```

Comment:
I have to say I wondered about this a bit when reviewing last week. Unpacking part of the script into a temporary script and calling it from the original "outer script" seemed like an interesting trick to me, but I had a hard time understanding *why* we are doing that. Off and on over the weekend I could not really come up with an answer that made me happy.

I discussed this with Markus earlier today and we just came to the conclusion that it probably has something to do with having one script name less to be concerned about while ensuring the correct pairing of "inner" and "outer" script between the two separate variants (local and aws). (?)

Maybe we want to add some good reasoning/commenting around the divider line. Just to ensure that someone coming across this code in 5 years has an easier way to accept that this is an interesting technique aiding the proper paring of two scripts and not for some deeper reason that needs to be deciphered.

Reply:
> I discussed this with Markus earlier today and we just came to the conclusion that it probably has something to do with having one script name less to be concerned about while ensuring the correct pairing of "inner" and "outer" script between the two separate variants (local and aws).

That's essentially it. I was trying to avoid the pattern seen in `tools/docker-go` where we have a somewhat generic "run this in Docker" script that's decoupled from the business logic of what to run. Such a script ends up having to support the superset of arguments for all its callers (tasks in `Makefile.toml`, and `gomod.rs` in `buildsys`), which is confusing to implement and also to use correctly (since `--help` just shows all the supported args).

The scripts could be slightly more opinionated and make the SDK image a required parameter. During development, I found it helpful to both the "host" and "container" execution modes available since the "aws" profile type needed a lot of debugging.

If I liked `docker-go` a bit better, or if I didn't care about preserving both modes of execution, then I'd probably have specialized scripts that called into a shared executor script.